core/src/main/java/org/apache/accumulo/core/summary/SummarySerializer.java

```diff
@@ -0,0 +1,543 @@
+        er |= lgs.exceedsRange(ke.getStartRow(), ke.getEndRow());
+        if (er) {
+          return er;
+        }
+      }
+    }
+
+    return er;
+  }
+
+  public boolean exceededMaxSize() {
+    return allSummaries == null;
+  }
+
+  private static class SummaryStoreImpl implements org.apache.accumulo.core.client.summary.Summarizer.StatisticConsumer {
+
+    HashMap<String,Long> summaries;
+
+    @Override
+    public void accept(String summary, long value) {
+      summaries.put(summary, value);
+    }
+  }
+
+  private static class LgBuilder {
+    private Summarizer summarizer;
+    private SummarizerConfiguration conf;
+    private Collector collector;
+
+    private int maxSummaries = 10;
+
+    private int cutoff = 1000;
+    private int count = 0;
+
+    private List<SummaryInfo> summaries = new ArrayList<>();
+
+    private Key lastKey;
+
+    private SummaryStoreImpl sci = new SummaryStoreImpl();
+
+    private String name;
+
+    private boolean sawFirst = false;
+    private Text firstRow;
+
+    private boolean finished = false;
+
+    public LgBuilder(SummarizerConfiguration conf, Summarizer kvs) {
+      this.conf = conf;
+      this.summarizer = kvs;
+      this.name = "<DEFAULT>";
+      this.collector = kvs.collector(conf);
+    }
+
+    public LgBuilder(SummarizerConfiguration conf, Summarizer kvs, String name) {
+      this.conf = conf;
+      this.summarizer = kvs;
+      this.name = name;
+      this.collector = kvs.collector(conf);
+    }
+
+    public void put(Key k, Value v) {
+      collector.accept(k, v);
+      count++;
+
+      if (!sawFirst) {
+        firstRow = k.getRow();
+        sawFirst = true;
+
+      }
+
+      if (count >= cutoff) {
+        sci.summaries = new HashMap<>();
+        collector.summarize(sci);
+        collector = summarizer.collector(conf);
+        addSummary(k.getRow(), sci.summaries, count);
+        count = 0;
+      }
+
+      lastKey = k;
+    }
+
+    private List<SummaryInfo> merge(int end) {
+      List<SummaryInfo> mergedSummaries = new ArrayList<>();
+      for (int i = 0; i < end; i += 2) {
+        int mergedCount = summaries.get(i).count + summaries.get(i + 1).count;
+        summarizer.combiner(conf).merge(summaries.get(i).summary, summaries.get(i + 1).summary);
+        mergedSummaries.add(new SummaryInfo(summaries.get(i + 1).getLastRow(), summaries.get(i).summary, mergedCount));
+      }
+      return mergedSummaries;
+    }
+
+    private void addSummary(Text row, Map<String,Long> summary, int count) {
+      Preconditions.checkState(!finished);
+      summaries.add(new SummaryInfo(row, summary, count));
+
+      if (summaries.size() % 2 == 0 && summaries.size() > maxSummaries) {
+        summaries = merge(summaries.size());
+        cutoff *= 2;
+      }
+    }
+
+    boolean collapse() {
+      Preconditions.checkState(finished);
+      if (summaries.size() <= 1) {
+        return false;
+      }
+
+      int end = summaries.size();
+      if (end % 2 == 1) {
```

Comment:
There lower bound is determined by the method that calls it.  That method keeps calling collapse as long as its greater than a certain size.